### PR TITLE
Feature add email resender

### PR DESCRIPTION
**Pull request by 24eeeee**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/400

### CODE REVIEW

File: src/UserService.Business/Helpers/Email/EmailResender.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                    try
+                    {
+                        IOperationResult<bool> respond = _rcSendEmail.GetResponse<IOperationResult<bool>>(_emailRequests[i]).Result.Message;
+                        if (respond.IsSuccess)
+                        {
+                            _emailRequests.RemoveAt(i);
+                        }
+                        else
+                        {
+                            _logger.LogWarning($"Can not send email. Email remain in resend queue.");
```
Comment (KostinP) [inline src/UserService.Business/Helpers/Email/EmailResender.cs:40]: What email were you trying to forward?

File: src/UserService.Business/Helpers/Email/EmailResender.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                        }
+                    }
+                    catch (Exception exc)
+                    {
+                        _logger.LogError(exc, $"Errors while sending email. Email remain in resend queue.");
```
Comment (KostinP) [inline src/UserService.Business/Helpers/Email/EmailResender.cs:45]: What is the error message?

File: src/UserService.Business/Helpers/Email/EmailResender.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.UserService.Business.Helpers.Email
+{
+    public class EmailResender
+    {
+        private static readonly IRequestClient<ISendEmailRequest> _rcSendEmail;
+        private static readonly ILogger<EmailResender> _logger;
```
Comment (KostinP) [inline src/UserService.Business/Helpers/Email/EmailResender.cs:14]: Add constructor (.NET will inject dependencies)

File: src/UserService/Startup.cs [hunk 4]
```diff
@@ -137,6 +140,12 @@ public Startup(IConfiguration configuration)
                 .GetSection(BaseRabbitMqConfig.SectionName)
                 .Get<RabbitMqConfig>();
 
+            _emailEngineConfig = Configuration
+                .GetSection(EmailEngineConfig.SectionName)
+                .Get<EmailEngineConfig>();
+
+            Task.Run(() => EmailResender.Start(_emailEngineConfig.ResendInterval));
```
Comment (KostinP) [inline src/UserService/Startup.cs:147]: Add read from enviroment

File: src/UserService/Startup.cs [hunk 5]
```diff
@@ -33,6 +35,7 @@ public class Startup : BaseApiInfo
 
         private readonly BaseServiceInfoConfig _serviceInfoConfig;
         private readonly RabbitMqConfig _rabbitMqConfig;
+        private readonly EmailEngineConfig _emailEngineConfig;
```
Comment (KostinP) [inline src/UserService/Startup.cs:38]: Why is not local variable?

**Review by KostinP — CHANGES_REQUESTED** [review]
Thank you for the courage in posting! This makes it easier to comment on the code.

We are also waiting:
- Update CommandTests
- Add EmailResenderTests

File: src/UserService/appsettings.json [hunk 6]
```diff
@@ -14,6 +14,9 @@
     "Id": "CBE144F5-AD53-4D0D-B6A1-39133E3F7D2D",
     "Name": "UserService"
   },
+  "EmailEngineConfig": {
+    "ResendInterval": 60
```
Comment (TIove) [inline src/UserService/appsettings.json:18]: It is better to use postfix "InMinutes" or "InSeconds", because now it is unclear

**Review by KostinP — APPROVED** [review]

**Review by TIove — APPROVED** [review]

File: src/UserService.Business/Commands/User/CreateUserCommand.cs [hunk 7]
```diff
@@ -87,28 +89,38 @@ private void SendEmail(DbUser dbUser, string password, List<string> errors)
...
                         email.Value,
                         templateLanguage,
-                        templateValues
-                       )).Result.Message;
+                        templateValues);
+                IOperationResult<bool> rcSendEmailResponse = _rcSendEmail
+                    .GetResponse<IOperationResult<bool>>(emailRequest)
+                    .Result
+                    .Message;
+
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/CreateUserCommand.cs:102]: remove empty line

**Review by MariyaSin — APPROVED** [review]

**Review by BelehovEgor — APPROVED** [review]

**Pull request merged**